cpp/src/arrow/compute/kernels/argsort.cc

```diff
@@ -0,0 +1,159 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/kernels/argsort.h"
+#include <algorithm>
+#include <numeric>
+#include <vector>
+#include "arrow/builder.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/expression.h"
+#include "arrow/compute/logical_type.h"
+#include "arrow/type_traits.h"
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+template <typename ArrayType>
+bool CompareValues(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.Value(lhs) < array.Value(rhs);
+}
+
+template <typename ArrayType>
+bool CompareViews(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.GetView(lhs) < array.GetView(rhs);
+}
+
+template <typename ArrowType, typename Compararor>
+class ArgsortKernelImpl : public ArgsortKernel {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+
+ private:
+  Compararor compare;
+
+  Status ArgsortImpl(FunctionContext* ctx, const std::shared_ptr<ArrayType>& values,
+                     std::shared_ptr<Array>* offsets) {
+    std::vector<uint64_t> ind(values->length());
+    std::iota(ind.begin(), ind.end(), 0);
+    auto nulls_begin = ind.end();
+
+    if (values->null_count()) {
+      nulls_begin =
+          std::stable_partition(ind.begin(), ind.end(),
+                                [&values](uint64_t ind) { return !values->IsNull(ind); });
+    }
+    std::stable_sort(ind.begin(), nulls_begin,
+                     [&values, this](uint64_t left, uint64_t right) {
+                       return compare(*values, left, right);
+                     });
+    UInt64Builder builder(ctx->memory_pool());
+    RETURN_NOT_OK(builder.AppendValues(ind.begin(), ind.end()));
```

Comment:
You're allocating double the memory that you need. Can you use `AllocateBuffer` instead to perform a single memory allocation, and sort in-place? Then you can construct `offsets` from that buffer